### PR TITLE
kconfig: Hide the DesignWare I2C driver on unsupported platforms

### PR DESCRIPTION
**Pull request by SebastianBoe**
When users are configuring nRF applications they are given the option
to enable the DesignWare I2C driver. But they should not be given this
option because nRF SoCs do not have DesignWare I2C HW.

This commit hides the driver by default by introducing the config
option HAS_I2C_DW.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

### CODE REVIEW

File: boards/arc/em_starterkit/Kconfig.board [hunk 1]
```diff
@@ -8,6 +8,7 @@ config BOARD_EM_STARTERKIT
 	bool "ARC EM Starter Kit"
 	depends on SOC_EMSK
 	select HAS_DTS
+	select HAS_I2C_DW
```
Comment (galak) [inline boards/arc/em_starterkit/Kconfig.board]: Seems like this should really be in SoC not board.

**Review by galak — CHANGES_REQUESTED** [review]
"select HAS_I2C_DW" should be happening in the SoC Kconfig files, not board ones.

File: drivers/i2c/Kconfig.dw [hunk 2]
```diff
@@ -4,11 +4,17 @@
...
 
+# This option signifies whether DesignWare I2C compatible HW is
+# available
+config HAS_I2C_DW
+	bool
+
 menuconfig I2C_DW
 	bool "Design Ware I2C support"
 	select HAS_DTS_I2C
+	depends on HAS_I2C_DW
```
Comment (anangl) [inline drivers/i2c/Kconfig.dw]: Wouldn't it be better to depend on SoC families instead? To stay consistent with other drivers' Kconfigs (see for instance esp32, nrfx, stm32, also SPI driver).

File: boards/arc/em_starterkit/Kconfig.board (hunk 1 above)
Comment (SebastianBoe) [reply boards/arc/em_starterkit/Kconfig.board]: Fixed.

**Comment by SebastianBoe** [comment]
> "select HAS_I2C_DW" should be happening in the SoC Kconfig files, not board ones.

Fixed.

@galak : Please re-assess the review when convenient.

File: drivers/i2c/Kconfig.dw (hunk 2 above)
Comment (SebastianBoe) [reply drivers/i2c/Kconfig.dw]: Discussed offline.

Doing it like this allows Kconfig.dw to be used with out-of-tree HW support.

**Comment by nashif** [comment]
I am not following why this driver specifically, this issue would be the same for all drivers. What is so special about this? Why the SPI DesignWare driver is not appearing for you?

Really do not like the introduction of HAS_ on drivers, we did have a lengthy discussion about this. It basically creating a precedent here which might lead to all drivers have this HAS_ configs, something we need to avoid. dependency hell.

**Comment by galak** [comment]
> I am not following why this driver specifically, this issue would be the same for all drivers. What is so special about this? Why the SPI DesignWare driver is not appearing for you?
> 
> Really do not like the introduction of HAS_ on drivers, we did have a lengthy discussion about this. It basically creating a precedent here which might lead to all drivers have this HAS_ configs, something we need to avoid. dependency hell.

I think it because other drivers have Kconfig depends like:

menuconfig I2C_NRFX
	depends on SOC_FAMILY_NRF

or

menuconfig I2C_QMSI
        bool "QMSI I2C driver"
        depends on QMSI

Where the I2C_DW doesn't have any such depends, and thus shows up as an option on all SoCs.

**Review by galak — APPROVED** [review]

**Comment by nashif** [comment]
> I think it because other drivers have Kconfig depends like:

so what about SPI DesignWare? is that going to be changed in this fashion too?

**Comment by SebastianBoe** [comment]
> so what about SPI DesignWare? is that going to be changed in this fashion too?

It should be. But IMHO configuration visibility issues should be fixed incrementally.

**Comment by SebastianBoe** [comment]
> so what about SPI DesignWare? is that going to be changed in this fashion too?

It should be. But IMHO configuration visibility issues should be fixed incrementally.

**Comment by galak** [comment]
> It should be. But IMHO configuration visibility issues should be fixed incrementally.

lets at least open an issue for the SPI driver.

**Comment by SebastianBoe** [comment]
> lets at least open an issue for the SPI driver.

Done.

https://github.com/zephyrproject-rtos/zephyr/issues/10825

**Comment by SebastianBoe** [comment]
@nashif : This PR is prone to bitrot.

So it should be merged unless there is more feedback that needs to be addressed.

**Pull request merged**